contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,591 @@
+    DISPLAY("# times waited on job ready: %u\n", stats.waitReady);
+    DISPLAY("# times waited on job compressed: %u\n", stats.waitCompressed);
+    DISPLAY("# times waited on job Write: %u\n\n", stats.waitWrite);
+}
+
+static int compressFilename(const char* const srcFilename, const char* const dstFilenameOrNull)
+{
+    BYTE* const src = malloc(FILE_CHUNK_SIZE);
+    unsigned const stdinUsed = !strcmp(srcFilename, stdinmark);
+    FILE* const srcFile = stdinUsed ? stdin : fopen(srcFilename, "rb");
+    const char* const outFilenameIntermediate = (stdinUsed && !dstFilenameOrNull) ? stdoutmark : dstFilenameOrNull;
+    const char* outFilename = outFilenameIntermediate;
+    char fileAndSuffix[MAX_PATH];
+    size_t const numJobs = MAX_NUM_JOBS;
+    int ret = 0;
+    adaptCCtx* ctx = NULL;
+    UTIL_getTime(&g_startTime);
+    g_streamedSize = 0;
+
+    if (!outFilenameIntermediate) {
+        if (snprintf(fileAndSuffix, MAX_PATH, "%s.zst", srcFilename) + 1 > MAX_PATH) {
+            DISPLAY("Error: output filename is too long\n");
+            ret = 1;
+            goto cleanup;
+        }
+        outFilename = fileAndSuffix;
+    }
+
+    /* checking for errors */
+    if (!srcFilename || !outFilename || !src || !srcFile) {
+        DISPLAY("Error: initial variables could not be allocated\n");
+        ret = 1;
+        goto cleanup;
+    }
+
+    /* creating context */
+    ctx = createCCtx(numJobs, outFilename);
+    if (ctx == NULL) {
+        ret = 1;
+        goto cleanup;
+    }
+
+    /* create output thread */
+    {
+        pthread_t out;
+        if (pthread_create(&out, NULL, &outputThread, ctx)) {
+            DISPLAY("Error: could not create output thread\n");
+            ret = 1;
+            goto cleanup;
+        }
+    }
+
+    /* create compression thread */
+    {
+        pthread_t compression;
+        if (pthread_create(&compression, NULL, &compressionThread, ctx)) {
+            DISPLAY("Error: could not create compression thread\n");
+            ret = 1;
+            goto cleanup;
+        }
+    }
+
+    /* creating jobs */
+    for ( ; ; ) {
+        size_t const readSize = fread(src, 1, FILE_CHUNK_SIZE, srcFile);
+        if (readSize != FILE_CHUNK_SIZE && !feof(srcFile)) {
+            DISPLAY("Error: problem occurred during read from src file\n");
+            ctx->threadError = 1;
+            ret = 1;
+            goto cleanup;
+        }
+        g_streamedSize += readSize;
+        /* reading was fine, now create the compression job */
+        {
+            int const error = createCompressionJob(ctx, src, readSize);
+            if (error != 0) {
+                ret = error;
+                ctx->threadError = 1;
+                goto cleanup;
+            }
+        }
+        if (feof(srcFile)) {
+            DEBUGLOG(2, "THE STREAM OF DATA ENDED %u\n", ctx->nextJobID);
+            ctx->lastJobID = ctx->nextJobID;
+            break;
+        }
+    }
+
+cleanup:
```

Comment:
One way to avoid `goto`,
is to create 2 nested functions,
the first one being in charge of creating / freeing resources,
while the second one just uses them.

This way, you can easily have multiple exit points within function 2,
and still get the guarantee that resources will be properly freed.